### ISSUE TITLE
[native-image] Too many files open error, many open fd pointing to /sys/fs/cgroup/memory/memory.limit_in_bytes

### ISSUE THREAD

**Issue by geek29** [oracle/graal#1494]
I am building simple http server based on netty plot demo (graalvm-demos/native-netty-plot) which demonstrates the use of isolates and image singletons. What I have found is that after a certain amount of iterations, running image keeps fds open for the following files

- /sys/fs/cgroup/memory/memory.limit_in_bytes
- /dev/urandom
- socket:[xxxxxxx]

Eventually, application starts to fail giving out `too many files open` error. 
As isolate is copy of JVM, I tried listing out open fds of equivalent java process which does keep a fd open to `/dev/urandom` but not `/sys/fs/cgroup/memory/memory.limit_in_bytes`

1. is this expected when running with isolates?
2. Will each isolate spawned keep fd open pointing to `/dev/urandom` ? How to ensure that it gets closed on isolate destroy?
3. Fd for file `/sys/fs/cgroup/memory/memory.limit_in_bytes` is a leak?

How can I make sure that after isolate is destroyed, above fds too get closed?

Steps to reproduce

1. Build native image of graalvm-demos/native-netty-plot and run. Note process ID
2. Run following script which lists out FDs 
3. Hit webbrowser many times manually and see open FDs

```
NODE_FD=`ls -l /proc/$1/fd | grep anon_inode | wc -l`
RANDOM_FD=`ls -l /proc/$1/fd | grep random | wc -l`
PIPE_FD=`ls -l /proc/$1/fd | grep pipe | wc -l`
CGROUP_FD=`ls -l /proc/$1/fd | grep cgroup | wc -l`
SOCK_FD=`ls -l /proc/$1/fd | grep socket | wc -l`
TOTAL=`ls -l /proc/$1/fd | wc -l`
echo "NODE_FD=$NODE_FD, RANDOM_FD=$RANDOM_FD, PIPE_FD=$PIPE_FD, CGROUP_FD=$CGROUP_FD, SOCK_FD=$SOCK_FD, TOTAL=$TOTAL"
```

**Comment by geek29** [comment]
Below is simple program which can reproduce the error. Type "exec 10" - it will spawn ten isolates. Inside each isolate, it creates an instance of `FileInputStream` and `SecureRandom` (which opens FileInputStream to file /dev/random). I close stream explicitely where as `SecureRandom` does not close it.
`SecureRandom` is created in my original program by SSL Socket layer.

My guess is that FDs remain open as Stream objects are never collected but simply destroyed as part of entire isolate destruction, hence never `finalize()` 'ed. Behaviour I see is that not only FD which is printed (hence closed in finally block) reappears again in list of open FDs (`ls -l /proc/${pid}/fd`) when same file is opened again in next isolate but new FDs pointing to same file also appear. Here isolates are created and destroyed in sequence but my original program its concurrent creation of isolation.

Does native-image isolate support such execution logic where many isolate (readOnly) access to same file? If not, is there any way where file-access can be shared between different isolates?
Whats the best way to ensure File descriptors (sockets etc) are properly closed and finalized?
Does ` Isolates.tearDownIsolate(renderingContext)`  call finalize method of all the files, streams etc open duruing isolate life-cycle? If not, does native-image isolate API provide any method to explicitely invoke it to on a reference?
How to debug leaks inside isolates? I tried heapdump but did not find anything inside it.



```

import org.graalvm.nativeimage.*;
import org.graalvm.nativeimage.c.function.CEntryPoint;
import org.graalvm.nativeimage.c.type.CCharPointer;
import org.graalvm.nativeimage.c.type.CTypeConversion;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.security.SecureRandom;
import java.util.Scanner;

public class FDTest {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        while(true) {
            String command = scanner.nextLine();
            String[] array = command.split(" ");
            switch (array[0]) {
                case "exec" :
                    exec(Integer.parseInt(array[1]));break;
            }
        }
    }

    private static void exec(int count) {
        IsolateRunner runner = new IsolateRunner();
        for(int i=0;i<count;i++) {
            runner.exec("" + 10, "");
        }
    }
}


class IsolateRunner {

    private static final Logger logger = LoggerFactory.getLogger(IsolateRunner.class);

    public void exec(String body, String path){
        IsolateThread renderingContext = Isolates.createIsolate(Isolates.CreateIsolateParameters.getDefault());
        IsolateThread nettyContext = CurrentIsolate.getCurrentThread();
        ObjectHandle bodyHandle = copyString(renderingContext, body);
        ObjectHandle pathHandle = copyString(renderingContext, path);
        ObjectHandle resultHandle = applicationEntryPoint(renderingContext, nettyContext, bodyHandle, pathHandle);
        logger.info("Application is complete");
        String result = ObjectHandles.getGlobal().get(resultHandle);
        ObjectHandles.getGlobal().destroy(resultHandle);
        logger.info("Handles destroyed");
        Isolates.tearDownIsolate(renderingContext);
        logger.info("Isolate destroyed");
    }

    private static ObjectHandle copyString(IsolateThread targetContext, String sourceString) {
        /* Convert the source Java string to a C string. */
        try (CTypeConversion.CCharPointerHolder cStringHolder = CTypeConversion.toCString(sourceString)) {
            /* Cross the isolate boundary with a C string as the parameter. */
            return copyString(targetContext, cStringHolder.get());
        }
    }

    @CEntryPoint
    private static ObjectHandle copyString(@CEntryPoint.IsolateThreadContext IsolateThread renderingContext, CCharPointer cString) {
        /* Convert the C string to the target Java string. */
        String targetString = CTypeConversion.toJavaString(cString);
        /* Encapsulate the target string in a handle that can be returned back to the source isolate. */
        return ObjectHandles.getGlobal().create(targetString);
    }

    // name of the pure random file (also used for setSeed())
    private static final String NAME_RANDOM = "/dev/random";
    // name of the pseudo random file
    private static final String NAME_URANDOM = "/dev/urandom";

    @CEntryPoint
    private static ObjectHandle applicationEntryPoint(@CEntryPoint.IsolateThreadContext IsolateThread renderingContext,
                                                    IsolateThread nettyContext, ObjectHandle bodyHandle, ObjectHandle pathHandle) {
        String body = ObjectHandles.getGlobal().get(bodyHandle);
        String path = ObjectHandles.getGlobal().get(pathHandle);
        ObjectHandles.getGlobal().destroy(bodyHandle);
        ObjectHandles.getGlobal().destroy(pathHandle);
        FileInputStream fis=null;
        try {
            SecureRandom secureRandom = new SecureRandom();
            Thread.sleep(Integer.parseInt(body));
            logger.info("Application gathered appResult " + secureRandom.getAlgorithm());
            fis = new FileInputStream(new File(NAME_RANDOM));
            fis = new FileInputStream(new File(NAME_RANDOM));
            logger.info("/DEV/RANDOM : File descriptor : " + fis.getFD());
            int fd1 = sun.misc.SharedSecrets.getJavaIOFileDescriptorAccess().get(fis.getFD());
            System.out.println("FD " + fis.getFD() + " Number=" + fd1);
            return copyString(nettyContext, ""+ System.currentTimeMillis());
        } catch (Exception e) {
            System.out.println("Error #1");
            e.printStackTrace();
            return copyString(nettyContext, e.getMessage());
        } finally {
            try {
                if(fis != null)
                    fis.close();
            } catch (IOException e) {
                System.out.println("fis.close :: Error #2");
                e.printStackTrace();
            }
            logger.info("runApplication complete");
        }
    }


}
```

**Comment by peter-hofer** [comment]
Thank you very much for your report @geek29!
1. I found that `/sys/fs/cgroup/memory/memory.limit_in_bytes` is leaked due to a bug and will prepare a fix.
1. `SecureRandom` creates a singleton `NativePRNG.RandomIO` instance that opens two file descriptors for `/dev/random` or `/dev/urandom` and never closes them. With a single isolate/VM, this does not have significant impact, but it becomes a conceptual problem when spawning several isolates. We'll discuss how to approach this.
1. native-image does not run finalizers, so code should not rely on them. Tearing down an isolate also does not trigger a garbage collection during which finalizers could run. This also means that `DirectByteBuffer` objects (from `ByteBuffer.allocateDirect`) potentially don't have their cleaner objects executed and leak their native memory unless they are explicitly closed. Resources can be explicitly released on tear-down by registering a tear-down hook in `RuntimeSupport.addTearDownHook()` however.

Regarding heap dumps, did you try the Java API example at https://www.graalvm.org/docs/reference-manual/native_heapdump/ ?

**Comment by geek29** [comment]
yes, I tried but heap dump that I took did not have any live isolates hence could not find anything inside it. I will try once again, probably through programmatically instead of using signals

**Comment by peter-hofer** [comment]
I fixed the `sysfs` file handle leak in https://github.com/oracle/graal/commit/ece874512d60352bbf462ab90cd123215b2ea7f7.

**Comment by geek29** [comment]
Is there any internal api which keeps a record of open fds which I can use to explicitly close in shutdownhook? As SecureRandom is JDK class I cant modify it. There could be another such fd leaks lurking anywhere. I have verified it that an open fd closed inside another isolate/same isolate explicitly avoids `Too many files open` error. The only issue I face is how to get hold of FD number of `/dev/random` or `/dev/urandom`

**Comment by peter-hofer** [comment]
We do not keep track of all open file descriptors. In principle, it can be done by inspecting all existing objects that might hold a file handle, like `java.io.FileDescriptor` and likely others, with heap-walking code similar to that used in `HeapDumpUtils`. I would consider this only as a crutch for diagnosing leaks however.
The right approach would be to address this leak in the `SecureRandom` implementation itself. While it is a JDK class, we have the (non-API) substitution mechanism with which we can modify those classes. See the `@com.oracle.svm.core.annotate.TargetClass` annotation for details and the various classes named `Target_...` for examples. So what we could do is to substitute the `sun.security.provider.NativePRNG.initIO()` method or `NativePRNG.RandomIO` constructor with a method that installs a tear-down hook that closes the file descriptor. Alternatively, maybe even easier, it is possible to use `@InjectAccessors` to intercept when the `NativePRNG.INSTANCE` field is set and install the tear-down hook then.

**Comment by peter-hofer** [comment]
I came up with a fix for the `SecureRandom` file handle leak that uses substitutions, it should be merged today or after the weekend.

**Comment by peter-hofer** [comment]
Fixed in 0840b5f0fdf3f70fef0e06c7a9a60f54142dc36c. Thanks for your report and helpful diagnosis!
